spqr/__init__.py

```diff
@@ -0,0 +1,349 @@
+        @param beta1: Tile width.
+        @param beta2: Tile height (see SpQR publication for more details).
+        @param dense_weights: The weights, first and second order data is stored in buff0. A single row tile
+        looks is stored in a 64-bit chunk as follows:
+
+                   |  w_s  | w_z | w_0 | ... | w_15 | c_0 | unused
+        bits(64):      3      3     3           3      8      2
+
+        w_s - first order quantized row scale
+        w_z - first order quantized row zero-point.
+
+        w_i - i-th quantized weight
+
+        c_0 - partial 8-bit chunk of the second order which the kernel recovers via warp shuffle reduction opeartion.
+
+        @param row_offsets: int32 CSR Row offsets if row_offsets.shape[0] == m + 1. PTCSR row offsets otherwise.
+        @param col_vals: int32 | values | column id | buffer.
+                                 16-bit    16-bit
+                                 higher     lower
+        @param in_perm: Optional input permutation (see SpQR publication for more details).
+        """
+        super().__init__()
+        self.m = rows
+        self.n = cols
+        self.bits = bits
+        self.beta1 = beta1
+        self.beta2 = beta2
+
+        self.dense_weights = nn.Parameter(dense_weights, requires_grad=False)
+        self.row_offsets = nn.Parameter(row_offsets, requires_grad=False)
+        self.col_vals = nn.Parameter(col_vals, requires_grad=False)
+        self.in_perm = nn.Parameter(in_perm, requires_grad=False)
+
+    @staticmethod
+    def create_placehodler(rows, cols, bits, beta1, beta2,
+                           dense_weights_shape: int,
+                           row_offsets_shape: int,
+                           col_vals_shape: int,
+                           in_perm_shape: int):
+        dense_weights = nn.Parameter(torch.empty(dense_weights_shape, dtype=torch.int64), requires_grad=False)
+        row_offsets = nn.Parameter(torch.empty(row_offsets_shape, dtype=torch.int32), requires_grad=False)
+        col_vals = nn.Parameter(torch.empty(col_vals_shape, dtype=torch.int32), requires_grad=False)
+        in_perm = nn.Parameter(torch.empty(in_perm_shape, dtype=torch.int64), requires_grad=False)
+
+        return QuantizedLinear(
+            rows, cols, bits, beta1, beta2,
+            dense_weights,
+            row_offsets,
+            col_vals,
+            in_perm
+        )
+
+    @staticmethod
+    def _calculate_weight_buffer_size(m, n, beta1, beta2):
+        total_blocks = updiv(m, beta1) * updiv(n, beta2)
+        per_tile_row = 1
+        block_size = beta1 * per_tile_row
+        return block_size * total_blocks
+
+    @staticmethod
+    def _allocate_weight_buffer(m, n, beta1, beta2, device) -> torch.Tensor:
+        sz0 = QuantizedLinear._calculate_weight_buffer_size(m, n, beta1, beta2)
+        return torch.zeros(sz0, dtype=torch.int64, device=device)
+
+    @staticmethod
+    def from_legacy(spqr_legacy: SPQRLegacy, model_args: ModelArgs, device):
```

Comment:
Is this needed only if we did not convert model from legacy format already  using convert_legacy_format.py?
What if the model is in HF format?

Reply:
Yes.
The HF format loading is not handled here - it's done on HF-side, not implemented in this repo.